src/UI/Tab.re

```diff
@@ -21,31 +21,92 @@ module Constants = {
   let minWidth = 125;
 };
 
-module Colors = {
-  let activeBackground = Theme.Colors.Tab.activeBackground;
-  let activeForeground = Theme.Colors.Tab.activeForeground;
-  let inactiveBackground = Theme.Colors.Tab.inactiveBackground;
-  let inactiveForeground = Theme.Colors.Tab.inactiveForeground;
-};
+module Colors = Theme.Colors.Tab;
 
-let proportion = p => float(Constants.minWidth) *. p |> int_of_float;
+let proportion = factor =>
+  float(Constants.minWidth) *. factor |> int_of_float;
 
 module Styles = {
   open Style;
 
-  let container = (~mode, ~isActive, ~showHighlight, ~theme) => {
-    let background =
-      theme#color(
-        isActive ? Colors.activeBackground : Colors.inactiveBackground,
-      );
-    let modeColor = Theme.Colors.Oni.backgroundFor(mode) |> theme#color;
-    let borderColor = isActive && showHighlight ? modeColor : background;
+  let container =
+      (
+        ~mode as _,
+        ~isGroupFocused,
+        ~isActive,
+        ~isHovered,
+        ~isModified,
+        ~theme,
+      ) => {
+    let background = {
+      let unhovered =
+        switch (isActive, isGroupFocused) {
+        | (false, _) => Colors.inactiveBackground
+        | (true, false) => Colors.unfocusedActiveBackground
+        | (true, true) => Colors.activeBackground
+        };
+
+      if (isHovered) {
+        theme#tryColor(
+          isGroupFocused
+            ? Colors.unfocusedHoverBackground : Colors.hoverBackground,
+        )
+        |> Option.value(~default=theme#color(unhovered));
+      } else {
+        theme#color(unhovered);
+      };
+    };
+
+    let borderTop = {
+      let color =
+        if (isActive) {
+          background;
+        } else {
+          theme#tryColor(
+            isGroupFocused
+              ? Colors.activeBorderTop : Colors.unfocusedActiveBorderTop,
+          )
+          |> Option.value(~default=background);
+        };
+
+      borderTop(~color, ~width=2);
+    };
+
+    let borderBottom = {
+      let color = {
+        let unhovered =
+          (
+            switch (isActive, isGroupFocused, isModified) {
+            | (false, _, false) => Colors.border
+            | (false, false, true) => Colors.unfocusedInactiveModifiedBorder
+            | (false, true, true) => Colors.inactiveModifiedBorder
+            | (true, false, true) => Colors.unfocusedActiveModifiedBorder
+            | (true, true, true) => Colors.activeModifiedBorder
+            | (true, false, false) => Colors.unfocusedActiveBorder
+            | (true, true, false) => Colors.activeBorder
+            }
+          )
+          |> theme#tryColor
+          |> Option.value(~default=background);
+
+        if (isHovered) {
+          (isGroupFocused ? Colors.unfocusedHoverBorder : Colors.hoverBorder)
+          |> theme#tryColor
+          |> Option.value(~default=unhovered);
+        } else {
+          unhovered;
+        };
+      };
+
+      borderBottom(~color, ~width=1);
+    };
```

Comment:
Did I mention that the vscode theming is insanely complex? Well, it's actually worse. In the UI code there's null checks everywhere special casing the absence of a theme color in various ways. The tab component alone is 1500 LOC, and the theming is scattered all over the place. Sometimes you have to look into several UI components to figure out the exact behaviour of a single color.

So these 30 lines of code above is basically just to figure out the color of the bottom border. I expect we might not want to be this rigorous everywhere though, and instead approximate by specifying sensible defaults instead of `null`/`unspecified` as vsode does. This is mostly just to make sure the API can handle it if we need to.